File: DSA/Merge_Sort.java
import java.util.Arrays;


public class Merge_Sort{


    static int[] mergeSort(int[] arr){

        if (arr.length==1) {
            return  arr;
        }

        int mid = arr.length/2;
        int[] left = mergeSort(Arrays.copyOfRange(arr, 0, mid));
        int[] right = mergeSort(Arrays.copyOfRange(arr, mid, arr.length));

        return merge(left, right);

    }


    static int[] merge(int[] first, int[] second){

        int[] mix = new int[first.length + second.length];
        int i = 0;
        int j = 0;
        int k = 0;


//  Comparing elements of the two arrays, less one is assigned to mix[] 

        while(i<first.length && j<second.length ){
            if(first[i] < second[j]){
                mix[k] = first[i];
                i++;
                k++;
            }
            else{
                mix[k] = second[j];
                j++;
                k++;
            }
        }


//  If one array elements are more than the other one, assign rest of array to mix[] 

        while(i<first.length){
            mix[k] = first[i];
            i++;
            k++;
        }

        while(j<second.length){
            mix[k] = second[j];
            j++;
            k++;
        }
        return mix;
    }





    public static void main(String[] args) {
        int[] arr = {8,4,3,12,5,6};
        arr=mergeSort(arr);
        System.out.println("After Merge sort : "+Arrays.toString(arr));
    }


}

File: DSA/MergeSort.c
/A program to sort an array using Merge Sort method/

#include <stdio.h>

void merge(int *array, int left, int mid, int right)
{
    int i,j,k;
    int n1=mid-left+1,n2=right-mid;
    int L[n1],R[n2];
    for(i=0;i<=n1-1;i++)
        L[i]=array[left+i];
    for(j=0;j<=n2-1;j++)
        R[j]=array[mid+1+j];
    i=0;
    j=0;
    k=left;
    while(i<=n1-1 && j<=n2-1)
    {
        if(L[i]<=R[j])
        {
            array[k]=L[i];
            i++;
        }
        else
        {
            array[k]=R[j];
            j++;
        }
        k++;
    }
    while(i<=n1-1)
    {
        array[k]=L[i];
        i++;
        k++;
    }
    while(j<=n2-1)
    {
        array[k]=R[j];
        j++;
        k++;
    }
}

void mergeSort(int *array, int left, int right)
{
    int mid;
    if(left<right)
    {
        mid=left+(right-left)/2;

        mergeSort(array,left,mid);
        mergeSort(array,mid+1,right);

        merge(array,left,mid,right);
    }
}

int main()
{
    int n,i;
	printf("Enter the size of the array: ");
	scanf("%d",&n);
	int arr[n];
	puts("\nEnter array elements...");
	for(i=0;i<=n-1;i++)
	    scanf("%d",&arr[i]);
	puts("\nBefore sorting...");
	for(i=0;i<=n-1;i++)
	    printf("%d\n",arr[i]);

    mergeSort(arr,0,n-1);

    puts("\nAfter sorting (Ascending Order)...");
	for(i=0;i<=n-1;i++)
	    printf("%d\n",arr[i]);
	puts("\nAfter sorting (Descending Order)...");
	for(i=n-1;i>=0;i--)
	    printf("%d\n",arr[i]);
        
    return 0;
}